moveit_experimental/moveit_jog_arm/src/jog_calcs.cpp

```diff
@@ -153,91 +163,96 @@ void JogCalcs::startMainLoop(JogArmShared& shared_variables, std::mutex& mutex)
     shared_variables.tf_moveit_to_cmd_frame = tf_moveit_to_cmd_frame_;
     mutex.unlock();
 
-    // Initialize the position filters to initial robot joints
-    while (!updateJoints() && ros::ok() && !stop_requested_)
+    // Only do jogging calculations if the robot should move, for efficiency
+    if (stop_requested_)
     {
-      mutex.lock();
-      incoming_joints_ = shared_variables.joints;
-      mutex.unlock();
-      ros::Duration(WHILE_LOOP_WAIT).sleep();
+      // If stopped, just keep the low-pass filters up to date with current joints so a jump doesn't occur when
+      // restarting
+      for (std::size_t i = 0; i < num_joints_; ++i)
+        position_filters_[i].reset(joint_state_.position[i]);
     }
-
-    // Prioritize cartesian jogging above joint jogging
-    if (!zero_cartesian_cmd_flag)
-    {
-      mutex.lock();
-      cartesian_deltas = shared_variables.command_deltas;
-      mutex.unlock();
-
-      if (!cartesianJogCalcs(cartesian_deltas, shared_variables, mutex))
-        continue;
-    }
-    else if (!zero_joint_cmd_flag)
+    else
     {
+      // Flag that incoming commands are all zero. May be used to skip calculations/publication
       mutex.lock();
-      joint_deltas = shared_variables.joint_command_deltas;
+      bool zero_cartesian_cmd_flag = shared_variables.zero_cartesian_cmd_flag;
+      bool zero_joint_cmd_flag = shared_variables.zero_joint_cmd_flag;
       mutex.unlock();
 
-      if (!jointJogCalcs(joint_deltas, shared_variables))
-        continue;
-    }
-    else
-    {
-      original_joint_state_ = joint_state_;
-      outgoing_command_ = composeOutgoingMessage(joint_state_);
-    }
-
-    // Halt if the command is stale or inputs are all zero, or commands were zero
-    mutex.lock();
-    bool stale_command = shared_variables.command_is_stale;
-    mutex.unlock();
-
-    if (stale_command || (zero_cartesian_cmd_flag && zero_joint_cmd_flag))
-    {
-      suddenHalt(outgoing_command_);
-      zero_cartesian_cmd_flag = true;
-      zero_joint_cmd_flag = true;
-    }
-
-    bool valid_nonzero_command = !zero_cartesian_cmd_flag || !zero_joint_cmd_flag;
-
-    // Send the newest target joints
-    if (!outgoing_command_.joint_names.empty())
-    {
-      mutex.lock();
-      // If everything normal, share the new traj to be published
-      if (valid_nonzero_command)
+      // Prioritize cartesian jogging above joint jogging
+      if (!zero_cartesian_cmd_flag)
       {
-        shared_variables.outgoing_command = outgoing_command_;
-        shared_variables.ok_to_publish = true;
+        mutex.lock();
+        cartesian_deltas = shared_variables.command_deltas;
```

Comment:
All these mutexes make the code unnecessarily hard to read. We should add thread safe accessor functions to shared_variables

Reply:
OK, but in a different PR, right?

Reply:
It should go a long way if you refactored JogArmShared into a class with mutexed setters and getters. That should ensure no race conditions can occur. Also note that some of the variables can just be converted into atomics.

Another thing I would recommend, is to use `std::lock_guard<foo>`. The thing about locking and unlocking is, that eventually you forget to unlock and there will be problems. Lock guard always unlocks the mutex at the end of scope so you can't really forget anything. You can also do this: 
```
std::lock_guard<std::mutex> lock(mutex)
{
  // Do things here
} // automatic unlocking.
```

In other words,, you set a temporary scope just to guide the lock_guard.

Reply:
I'll work on this in a followup PR. This one's already pretty big

Reply:
I'm going to approve this so we can move on to "fix mutexes" state :)